doc/CMakeLists.txt

```diff
@@ -0,0 +1,32 @@
+# Copyright (c) 2024-present The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or https://opensource.org/license/mit/.
+
+find_package(Doxygen COMPONENTS dot)
+
+if(DOXYGEN_FOUND)
+  set(doxyfile ${CMAKE_CURRENT_BINARY_DIR}/Doxyfile)
+  configure_file(Doxyfile.in ${doxyfile})
+
+  # In CMake 3.27, The FindDoxygen module's doxygen_add_docs()
+  # command gained a CONFIG_FILE option to specify a custom doxygen
+  # configuration file.
+  # TODO: Consider using it.
+  set(stampfile ${CMAKE_CURRENT_BINARY_DIR}/.stamp)
+  add_custom_command(
+    OUTPUT ${stampfile}
```

Comment:
Why not use an actual file that is produced by doxygen, like `OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/doxygen/html/index.html`?

Reply:
I've just reproduced the behaviour from the master branch.

Reply:
My experiments show that the behavior in `master` is to rerun `doxygen` every time `gmake docs` is executed whereas with this cmake approach `doxygen` is not rerun. To make experimenting easier I replaced the `doxygen` command with a shell script that only prints its arguments (preceded by `====` so that I can distinguish it easily in the output).

So, in `master`, running `gmake docs` for the first time produces:
```
/usr/local/bin/gmkdir -p doc/doxygen
/home/vd/bin/doxygen doc/Doxyfile FORCE
===== doxygen doc/Doxyfile FORCE
touch doc/doxygen/.stamp
```
and rerunning `gmake docs` immediately afterwards produces the same output. Thus, `doxygen` is re-executed. I also spot a bug that a second argument `FORCE` is passed to `doxygen` which it seems to ignore.

With this PR, running `cmake --build "[BUILD]" --verbose --target docs` for the first time produces (source and build directories replaced with `[SRC]` and `[BUILD]` for clarity):
```
Generating developer documentation
cd [SRC] && doxygen [BUILD]/doc/Doxyfile
===== doxygen [BUILD]/doc/Doxyfile
cd [SRC] && /usr/local/bin/cmake -E touch [BUILD]/doc/.stamp
gmake[3]: Leaving directory '[BUILD]'
Built target docs
```
and rerunning it a second time:
```
gmake[3]: Nothing to be done for 'doc/CMakeFiles/docs.dir/build'.
```
`doxygen` is not re-executed. This is the same with `Unix Makefiles` and with `Ninja`.

Consider the following:

```diff
   # TODO: Consider using it.
-  set(stampfile ${CMAKE_CURRENT_BINARY_DIR}/.stamp)
-  add_custom_command(
-    OUTPUT ${stampfile}
-    COMMAND Doxygen::doxygen ${doxyfile}
-    COMMAND ${CMAKE_COMMAND} -E touch ${stampfile}
-    WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}
-    DEPENDS ${doxyfile}
-    COMMENT "Generating developer documentation"
-    VERBATIM
-  )
   add_custom_target(docs
-    DEPENDS ${stampfile}
+    Doxygen::doxygen ${doxyfile}
+    COMMENT "Generating developer documentation"
+    WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}
   )
```
it is simpler and does like in `master` (except the bogus second argument to `doxygen`).

Reply:
Thanks! [Reworked](https://github.com/hebasto/bitcoin/pull/165#issuecomment-2080404349).